### ISSUE TITLE
Question: interactions erroneously reversed in some cases?

### ISSUE THREAD

**Issue by sjfleming** [saezlab/omnipath#20]
Hello!  I am new-ish to this kind of thing, but when I run
```
squidpy.gr.ligrec()
```
from `squidpy` (https://squidpy.readthedocs.io/en/stable/api/squidpy.gr.ligrec.html), it says it pulls information from `omnipath`.

I get an output metadata file that has the following two entries:
![image](https://user-images.githubusercontent.com/10214815/163245291-dbacd3be-1c3d-4405-8ddc-377e181d26d7.png)
and
![image](https://user-images.githubusercontent.com/10214815/163245739-f63ac055-d01f-4005-9962-e01f6915a500.png)

**These entries also seem to be present in the `omnipath` database via a direct python query.**

## My question:
Is it a bug that this interaction seems to be present in both "directions"?  The entry saying that TGFBR3 is a ligand and that TGFB1 is the receptor?

It gives "CellPhoneDB" as one of the sources for both of these entries.  But CellPhoneDB only has the interaction in one direction: with TGFBR3 as the "receptor".

There are other examples of interactions with "ligands" that explicitly have "receptor" in the gene name (though I know sometimes it's a bit more complicated...), and I am wondering if some of these might be in error.

**Comment by deeenes** [comment]
Hi,

That's a good question, indeed first it sounds weird that a receptor stimulates a ligand. In our [database building software](https://github.com/saezlab/pypath) `pypath` the interaction looks like this:

```
from pypath import omnipath
o = omnipath.db.get_db('omnipath')
o.interaction('TGFB1', 'TGFBR3').direction
{
    'undirected': <Evidences: CellPhoneDB, DIP, HPRD, Lit-BM-17, NetPath, SPIKE, Wang (6 references)>,
    (<Entity: TGFB1>, <Entity: TGFBR3>): <Evidences: Baccin2019, CellTalkDB, Cellinker, DLRP, EMBRACE, Fantom5, Guide2Pharma, HPRD, LRdb, Ramilowski2015, STRING, UniProt, connectomeDB2020, iTALK, talklr (1 references)>,
    (<Entity: TGFBR3>, <Entity: TGFB1>): <Evidences: Wang (0 references)>
}
```

As we see, only the resource _Wang_ supports the TGFBR3->TGFB1 direction, and without any literature reference. Checking [the original file ](https://www.researchgate.net/profile/Edwin-Wang-6/publication/264488606_Human_Signaling_Network_Version_6/data/53e102ed0cf2235f35271ea3/HuamnSignalingNet-v6.csv?origin=publication_detail) we find this interaction in line 60296:

```
grep -n TGFBR3  HuamnSignalingNet_v6.csv
...
60296:7049,TGFBR3,7040,TGFB1,Pos
```

One problem with _Wang_ is that it is a huge dataset from the lab of [Prof. Edwin Wang, ](http://www.cancer-systemsbiology.org/) but without literature references and detailed curation methods. They only point to their own five papers (Ma'ayan 2005, Awan 2007, Cui 2007, Li 2012, Newman 2013), which provide only short description of the construction of this network: [basically they started off from BioCarta pathway diagrams and Cancer Cell Map](https://omnipathdb.org/info#Cui2007), two old, already disbanded databases. And combined this with [their own literature curation](https://omnipathdb.org/info#CA1), and possibly later they added data from other sources too.

Trying to trace back the interaction TGFBR3 --(+)--> TGFB1, the earliest source where it can be found is [Cui et al. 2007](https://www.embopress.org/doi/full/10.1038/msb4100200) (Supp Table 9, row 4974). 

Taking a closer look at TGFBR3 finally answers the question. As UniProt writes: [Binds to TGF-beta. Could be involved in capturing and retaining TGF-beta for presentation to the signaling receptors.](https://www.uniprot.org/uniprot/Q03167) TGFBR3 is in fact not a receptor, but a proteoglycan facilitating the binding of TGFB1 to its real receptors, TGFBR1 and TGFBR2, by keeping these ligands anchored to itself: https://rupress.org/jcb/article/109/6/3137/29038/Membrane-anchored-and-soluble-forms-of-betaglycan, see also [here in the 5th paragraph](https://sci-hub.se/10.1056/NEJM199411103311907). This latter is the paper that other resources cite describing the opposite direction interaction!

First I thought _Wang et al._ is wrong, but seems like they were the only ones who got it right, in contrast to the ~7 other databases which, likely due to superficial curation, opted for TGFB1->TGFBR3. 

Still, in the context of squidpy this still might be a "wrong" interaction. One problem is how can we distinguish between this kind of interactions and conventional ligand-receptor links. Another issue is that the cell-cell communication prediction algorithms may not be ready to consider this kind of relationship. I would have expected CellPhoneDB to at least describe TGFBR3 as a co-receptor of TGFBR1-2, which would be an appropriate representation, but I see only simple interactions with five TGF type ligands: https://www.cellphonedb.org/fast-query/ligands-from-receptor?species=HOMO_SAPIENS&receptor=TGFBR3

Best,

Denes

**Comment by sjfleming** [comment]
Hi Denes,

Thank you so much for this comprehensive answer!  I really appreciate it.  The code snippet for pulling that info from `pypath` is also very helpful.  Also very helpful that you tracked down the truth about this interaction!  :)

It does seem like maybe this is a `squidpy` issue and not an `omnipath` issue... I see above that the `pypath` interaction is "undirected" from CellPhoneDB, but in my `squidpy` output I see it in both directions.  Maybe this is `squidpy`'s way of saying it's undirected... I'm not sure yet.

One last question though: the fact that the interaction is documented as being "undirected" from CellPhoneDB (via the `pypath` query)... is that in agreement with CellPhoneDB's database?  (I have a hard time navigating CellPhoneDB's database.) But it seems like they do not say the interaction (`CPI-SS085EE60B1`) is "undirected", but rather that "partner_a" is `P01137`, or TGFB1_HUMAN, and that "partner_b" is `Q03167`, or TGFBR3_HUMAN.  Then they also say that TGFB1 is `FALSE` for being a "receptor", while TGFBR3 is `TRUE` for being a "receptor".  This, if I'm reading it right, seems to say that the CellPhoneDB evidence should suppor the `(<Entity: TGFB1>, <Entity: TGFBR3>)` interaction, rather than `'undirected'`, in the `pypath` query.  Am I right about that?

Thanks!
Stephen

**Comment by deeenes** [comment]
Hi Stephen,

Above you see that in `pypath` this pair of proteins have three records by direction: undirected, TGFB1->TGFBR3 and TGFBR3-> TGFB1. In OmniPath this is [collapsed to two records,](https://omnipathdb.org/interactions?types=post_translational&partners=TGFBR3,TGFB1&fields=sources,references&source_target=AND&genesymbols=yes) because we don't show the undirected one if there is direction information. From the `pypath` representation it's also visible that the record from CellPhoneDB is undirected.

Looking into the half-processed CellPhoneDB data, we see that this interaction is labelled as "ligand-ligand":
```
from pypath.inputs import cellphonedb
c = cellphonedb.cellphonedb_interactions()
[i for i in c if i.id_a == 'P01137' and i.id_b == 'Q03167']
# [CellphonedbInteraction(id_a='P01137', id_b='Q03167', sources='CellPhoneDB', references='', interaction_type='ligand-ligand', type_a='ligand', type_b='ligand')]
```

We find the interaction at [line 950 in the raw CPDB data](https://github.com/Teichlab/cellphonedb/blob/master/cellphonedb/src/core/data/interaction_input.csv#L950); in the proteins table, TGFB1 is [secreted but not receptor,](https://github.com/Teichlab/cellphonedb/blob/master/cellphonedb/src/core/data/protein_input.csv#L120) and TGFBR3 is also [secreted but is a receptor](https://github.com/Teichlab/cellphonedb/blob/master/cellphonedb/src/core/data/protein_input.csv#L637). We can see the same in `pypath`:

```
from pypath.inputs import cellphonedb
ca = cellphonedb.cellphonedb_protein_annotations()
ca['Q03167'] # TGFBR3
# CellPhoneDBAnnotation(receptor=True, receptor_class=('receptor',), peripheral=False, secreted=True, secreted_class=('secreted',), transmembrane=True, integrin=False)
ca['P01137'] # TGFB1
# CellPhoneDBAnnotation(receptor=False, receptor_class=(), peripheral=False, secreted=True, secreted_class=('growthfactor',), transmembrane=False, integrin=False)
```

CellPhoneDB has no explicit directions and ligand annotations, hence in `pypath` we consider ligands all [secreted or non-receptor membrane bound proteins](https://github.com/saezlab/pypath/blob/cf9838f3edec29562bf2ae2443dde12bb35a38dd/pypath/inputs/cellphonedb.py#L75). Then, [ligands have priority over receptors](https://github.com/saezlab/pypath/blob/cf9838f3edec29562bf2ae2443dde12bb35a38dd/pypath/inputs/cellphonedb.py#L258), which means all secreted receptor become ligands. This might be not the best procedure, something that we could change, however then it would result ambiguous records. Finally, from CellPhoneDB [only "ligand-receptor" interactions are considered to be directed](https://github.com/saezlab/pypath/blob/cf9838f3edec29562bf2ae2443dde12bb35a38dd/pypath/resources/data_formats.py#L2826).

These details are faithfully represented in the [OmniPath annotations data](https://omnipathdb.org/annotations?resources=CellPhoneDB&proteins=TGFBR3). In the [intercell database](https://omnipathdb.org/intercell?resources=CellPhoneDB&proteins=TGFBR3), we also classify it as ligand (due to the above detailed reasons), however the `consensus_score` of the ligand record is very low (1), which means that the [quality filters in `OmnipathR`](https://saezlab.github.io/OmnipathR/reference/filter_intercell_network.html) would remove this wrong record. This quality filter is not implemented yet in the `omnipath` Python module, but it would be really useful to have it (I will open an issue for that).

In conclusion, we see that this is a kind of error, but probably a handful of proteins are affected (secreted receptors) in CellPhoneDB. I am wondering what could be a better representation for this? Maybe I should change `pypath.inputs.cellphonedb` in a way that receptors can not be ligands.

> It does seem like maybe this is a squidpy issue and not an omnipath issue... I see above that the pypath interaction is "undirected" from CellPhoneDB, but in my squidpy output I see it in both directions. Maybe this is squidpy's way of saying it's undirected... I'm not sure yet.

Actually this is what we see in OmniPath, a bi-directional interaction, as it is based on all the databases, not only CellPhoneDB.

Best,

Denes

**Comment by sjfleming** [comment]
Hi Denes,

Thanks again!  Now I think I understand.  Yes, I see how secreted receptors in CellPhoneDB are causing this to happen.  CellPhoneDB says `secreted`, and then OmniPath says that `secreted` implies `ligand`, and then `ligand` takes precedence over `receptor`.  Then, since the interaction ends up being ligand-ligand, it is "undirected" as far as OmniPath is concerned.

[As you mentioned, these cases are likely cases where CellPhoneDB's annotation might not be correct in the first place (as Wang might be right about the directionality here after all).]

The only thing I was still confused about was, if `omnipath` represents this interaction from CellPhoneDB as "undirected" due to the fact that it was classified as "ligand-ligand"... then why was I seeing the interaction at all in my `squidpy` analysis (for which I only used CellPhoneDB as a source)?  ... since, as you say, "we don't show the undirected one if there is direction information".  I thought maybe the undirected CellPhoneDB interaction would get dropped entirely.  But now I see what happens from your link ["collapsed to two records"](https://omnipathdb.org/interactions?types=post_translational&partners=TGFBR3,TGFB1&fields=sources,references&source_target=AND&genesymbols=yes): rather than being dropped, CellPhoneDB is now listed as evidence for both of these directions.

It all makes sense now!

Whether this is what "should" be happening or not... I do not know.  That is beyond my expertise, as someone who doesn't know too much about these kinds of interactions.  But either way, at least I understand the current logic.

Thanks again,
Stephen

**Comment by deeenes** [comment]
Hi Stephen,

Yes, you are completely right: _rather than being dropped, CellPhoneDB is now listed as evidence for both of these directions._ Citing again the complete `pypath` representation:

```
from pypath import omnipath
o = omnipath.db.get_db('omnipath')
o.interaction('TGFB1', 'TGFBR3').direction
{
    'undirected': <Evidences: CellPhoneDB, DIP, HPRD, Lit-BM-17, NetPath, SPIKE, Wang (6 references)>,
    (<Entity: TGFB1>, <Entity: TGFBR3>): <Evidences: Baccin2019, CellTalkDB, Cellinker, DLRP, EMBRACE, Fantom5, Guide2Pharma, HPRD, LRdb, Ramilowski2015, STRING, UniProt, connectomeDB2020, iTALK, talklr (1 references)>,
    (<Entity: TGFBR3>, <Entity: TGFB1>): <Evidences: Wang (0 references)>
}
```

There are at least 11 other databases (Baccin2019, CellTalkDB, Cellinker, EMBRACE,  Guide2Pharma, HPRD, LRdb, Ramilowski2015, connectomeDB2020, iTALK, talklr) that support one direction and one (Wang) which supports the other. Further 5 reporting it as undirected (CellPhoneDB, DIP, Lit-BM-17, NetPath, SPIKE) are [shown at both directions](https://omnipathdb.org/interactions?types=post_translational&partners=TGFBR3,TGFB1&fields=sources,references&source_target=AND&genesymbols=yes) in OmniPath. In OmniPath we merge the information, and the web server is not able to recover all the resource level details, e.g. if you query only for CellPhoneDB, then all interactions will be supported by CellPhoneDB, but the directions and effects might come from other resources. This is a current technical limitation of OmniPath, which we will address soon by moving to a much more powerful, graph DBMS based backend.

One more point: actually OmniPath does not return undirected interactions by default, but only with the `directed=no` parameter.

Best,

Denes

**Comment by grst** [comment]
What was your rationale of making every secreted receptor a ligand? Isn't basically every cell surface receptor secreted? 

Above, I could only find
> [...] however then it would result ambiguous records.

**Comment by grst** [comment]
Hi @deeenes, 

thanks for the extensive explanation! 

A tutorial with best practice recommendations on how to use omnipathdb for single-cell-based cell2cell communication analysis would be really helpful. I think there's also some relevant information in this issue: https://github.com/saezlab/pypath/issues/159? 

Cheers, 
Gregor
